shutdown: Destroy kernel last, make test shutdown order consistent

sedited opened the pull request:
The destruction/resetting of node context members in the tests should roughly follow the behavior of the `Shutdown` function in `init.cpp`.

This was originally requested by MarcoFalke in this [comment](https://github.com/bitcoin/bitcoin/pull/25065#discussion_r890161249) in response to the [original pull request](https://github.com/bitcoin/bitcoin/pull/25065) introducing the `kernel::Context`.

ryanofsky commented on src/test/util/setup_common.cpp:
```diff
@@ -144,6 +144,7 @@ BasicTestingSetup::BasicTestingSetup(const ChainType chainType, const std::vecto
 
 BasicTestingSetup::~BasicTestingSetup()
 {
+    m_node.kernel.reset();
```
In commit "tests: Reset node context members on ~ChainTestingSetup" (960112dbb1fb028b73b6e83932dd457f5a69ab5d)

I think this change makes sense but the log message is misleading because in the Shutdown() function the kernel context is currently destroyed _before_ chainman, mempool, scheduler, not after:

https://github.com/bitcoin/bitcoin/blob/d18a8f6f6969487021b8fd50391a2a8d1dc29844/src/init.cpp#L346-L350

The test code is more correct than the Shutdown() code because ~BasicTestingSetup runs after ~ChainTestingSetup, and the chainman and scheduler objects should be destroyed before kernel context because they might rely on it.

It'd be good to either clarify the log message in this commit to say mention this part of the change doesn't match Shutdown function, or update the Shutdown function to destroy the kernel context last like the test. (I could also update the Shutdown function in #28051 if you want this PR to be test-only)

ryanofsky reviewed (approved):
Code review ACK 960112dbb1fb028b73b6e83932dd457f5a69ab5d

TheCharlatan commented:
Yes.

TheCharlatan commented:
Updated 960112dbb1fb028b73b6e83932dd457f5a69ab5d -> 99a668ceb80b8241c9eb24dca2858dbdd3ddf7ee ([testSetupDestructorOrder_0](https://github.com/TheCharlatan/bitcoin/tree/testSetupDestructorOrder_0) -> [testSetupDestructorOrder_1](https://github.com/TheCharlatan/bitcoin/tree/testSetupDestructorOrder_1),  [compare](https://github.com/TheCharlatan/bitcoin/compare/testSetupDestructorOrder_0..testSetupDestructorOrder_1))

* Addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/28224#discussion_r1341571120), ammending commit to adjust the reset order in the Shutdown function.

ryanofsky reviewed (approved):
Code review ACK 99a668ceb80b8241c9eb24dca2858dbdd3ddf7ee

Changes look good, but I would suggest changing the PR title to something like: "shutdown: destroy kernel last and make test shutdown order consistent" because the current title "tests: Reset node context members on ~ChainTestingSetup" makes it seems like this is a test-only change.

Also think it would be better to split changes in `init.cpp` and `setup_common.cpp` into separate commits. The changes are both related to shutdown but they have different rationales and effects, so it doesn't really make sense for them to be in the same commit.

sedited committed:
shutdown: Destroy kernel last

Currently the shutdown function resets the kernel before the
chainman and scheduler. Invert this order by resetting the kernel
last, since they might rely on the kernel.

sedited committed:
tests: Reset node context members on ~BasicTestingSetup

The destruction/resetting of node context members in the tests should
roughly follow the behaviour of the Shutdown function in `init.cpp`.

TheCharlatan commented:
Updated 99a668ceb80b8241c9eb24dca2858dbdd3ddf7ee -> c1144f0076339c775f41d4b5fcfdc72191440d96 ([testSetupDestructorOrder_1](https://github.com/TheCharlatan/bitcoin/tree/testSetupDestructorOrder_1) -> [testSetupDestructorOrder_2](https://github.com/TheCharlatan/bitcoin/tree/testSetupDestructorOrder_2),  [compare](https://github.com/TheCharlatan/bitcoin/compare/testSetupDestructorOrder_2..testSetupDestructorOrder_2))

* No code change, just addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/28224#pullrequestreview-1692709982), splitting up the commit.

ryanofsky reviewed (approved):
Code review ACK c1144f0076339c775f41d4b5fcfdc72191440d96. No code changes since last review, just updated commits and descriptions

achow101 commented:
ACK c1144f0076339c775f41d4b5fcfdc72191440d96
